# Thesis
This repository includes code generated during my PhD time. Includes scripts for running the genome-resolved metagenomics workflow, script for annotation of butyrate producers and other parser scripts.

## Installation
All the software needed to run the genome-resolved metagenomics workflow can be installed via the conda package manager. Databases have to be installed separately based on the program used, and the path to the database should be updated accordingly in the scripts.

## Example
For testing the workflow, I have used the Zymo mock community from the following publication.
Rausch, P., Rühlemann, M., Hermes, B. M., Doms, S., Dagan, T., Dierking, K., Domin, H., Fraune, S., Von Frieling, J., Hentschel, U., Heinsen, F.-A., Höppner, M., Jahn, M. T., Jaspers, C., Kissoyan, K. A. B., Langfeldt, D., Rehman, A., Reusch, T. B. H., Roeder, T., … Baines, J. F. (2019). Comparative analysis of amplicon and metagenomic sequencing methods reveals key features in the evolution of animal metaorganisms. Microbiome, 7(1), 133. https://doi.org/10.1186/s40168-019-0743-1
